src/ExifViewer.jl

```diff
@@ -1,6 +1,132 @@
+Dict{Any, Any} with 2 entries:
+  "EXIF_TAG_FLASH_PIX_VERSION" => "FlashPix Version 1.0"
+  "EXIF_TAG_ORIENTATION"       => "Top-left"
+
+julia> read_tags(filepath; tags=[LibExif.EXIF_TAG_FLASH_PIX_VERSION, LibExif.EXIF_TAG_ORIENTATION])
+Dict{Any, Any} with 2 entries:
+    "EXIF_TAG_FLASH_PIX_VERSION" => "FlashPix Version 1.0"
+    "EXIF_TAG_ORIENTATION"       => "Top-left"
+
+julia> file = open(filepath, "r")
+julia> read_tags(file, tags=[LibExif.EXIF_TAG_FLASH_PIX_VERSION, LibExif.EXIF_TAG_ORIENTATION])
+Dict{Any, Any} with 2 entries:
+      "EXIF_TAG_FLASH_PIX_VERSION" => "FlashPix Version 1.0"
+      "EXIF_TAG_ORIENTATION"       => "Top-left"
+    
+```
+"""
+function read_tags(
+    data::Vector{UInt8};
+    ifds::Union{Int,NTuple,UnitRange} = 1:5,
+    read_all = false,
+    tags::Vector = [],
+    extract_thumbnail = false,
+)
+    ed_ptr = LibExif.exif_data_new_from_data(data, length(data))
+    if (ed_ptr == C_NULL)
+        return error("Unable to read EXIF data: invalid pointer")
+    end
+
+    ed = unsafe_load(ed_ptr)
+
+    ifds = collect(ifds)
+    checkbounds(Bool, collect(1:numifds(ed)), ifds) || throw(BoundsError(collect(1:numifds(ed)), ifds))
+    # ifds = filter(x -> (x > 0 && x <= 5), ifds)
+
+    @assert typeof(tags) == Vector{LibExif.ExifTag} || typeof(tags) == Vector{Any}
+    tags = read_all ? tags : Set(tags)
+    result = Dict{String,Any}()
+    str = Vector{Cuchar}(undef, 1024)
+
+    for i in ifds
+        content_ptr = ed.ifd[i] # ques: do we need to unref these too?
+        if (content_ptr == C_NULL)
+            return error("Unable to read IFD:", i)
+        end
+        data = unsafe_load(content_ptr)
+        if data.count == 0
+            continue
+        end
+        res = unsafe_wrap(Array, data.entries, data.count)
+        for i = 1:data.count
+            entry = unsafe_load(res[i])
+            condition = read_all ? read_all : entry.tag in tags
+            if condition
+                LibExif.exif_entry_get_value(Ref(entry), str, length(str))
+                tag = String(copy(str))[1:findfirst(iszero, str)-1]
+                tag = fixformat(tag, entry.format)
+                if string(entry.tag) ∉ keys(result)
+                    result[string(entry.tag)] = tag
+                end
+            end
+            if read_all == false
+                delete!(tags, entry.tag)
+                if tags == Set()
+                    break
+                end
+            end
+        end
+    end
+
+    # not sure we should include this
+    # if isempty(tags) != true
+    #     @info "Non-Existent Tags:" tags
+    # end
+
+    if (extract_thumbnail == true)
+        thumbnail_size = Int(ed.size)
+        thumbnail_data = unsafe_wrap(Array, ed.data, thumbnail_size)
+        result["EXIF_TAG_THUMBNAIL_DATA"] = thumbnail_data
+    end
+
+    LibExif.exif_data_unref(ed_ptr)
+    return result
+end
+
+read_tags(filepath::AbstractString; kwargs...) = read_tags(read(filepath); kwargs...)
```

Comment:
discussion in TiffImages https://github.com/tlnagy/TiffImages.jl/pull/79#discussion_r880478304) seems very interesting which will be a work in progress in understanding, I think I'll just finish this for now

Reply:
should I do this for io too?ensuring that `IO` is closed too

Reply:
> should I do this for io too? Ensuring that IO is closed too

Ensuring the resource is properly closed when not used anymore is a basic requirement unless you have special reasons to not do so. IO, memory, windows are such kind of resources. That is, we should always use `open(file) do ... end` syntax unless there are special reasons.

The issue tim and I discussed there is because `mmap` is a special usage -- I'm not very convinced that it's a good design, but it doesn't hurt much (since mmap is an advanced feature and most developers who know `mmap` have the skill to identify the resource issue)